ddsc/core/remotestore.py

```diff
@@ -85,27 +93,72 @@ def lookup_user_by_name(self, full_name):
         results = json_data['results']
         found_cnt = len(results)
         if found_cnt == 0:
-            raise ValueError("User not found:" + full_name)
+            raise NotFoundError("User not found:" + full_name)
         elif found_cnt > 1:
             raise ValueError("Multiple users with name:" + full_name)
         user = RemoteUser(results[0])
         if user.full_name.lower() != full_name.lower():
-            raise ValueError("User not found:" + full_name)
+            raise NotFoundError("User not found:" + full_name)
         return user
 
     def lookup_user_by_username(self, username):
         """
         Finds the single user who has this username or raises ValueError.
         :param username: str username we are looking for
-        :return: RemoteUser user we found
+        :return: RemoteUser: user we found
         """
         matches = [user for user in self.fetch_all_users() if user.username == username]
         if not matches:
-            raise ValueError('Username not found: {}.'.format(username))
+            raise NotFoundError('Username not found: {}.'.format(username))
         if len(matches) > 1:
             raise ValueError('Multiple users with same username found: {}.'.format(username))
         return matches[0]
 
+    def get_or_register_user_by_username(self, username):
+        """
+        Try to lookup user by username. If not found try registering the user.
+        :param username: str: username to lookup
+        :return: RemoteUser: user we found
+        """
+        try:
+            return self.lookup_user_by_username(username)
+        except NotFoundError:
+            return self.register_user_by_username(username)
+
+    def register_user_by_username(self, username):
+        """
+        Tries to register user with the first non-deprecated auth provider.
+        Raises ValueError if the data service doesn't have any default providers.
+        :param username: str: netid of the user we are trying to register
+        :return: RemoteUser: user that was created for our netid
+        """
+        default_providers = [prov.id for prov in self.get_auth_providers() if not prov.is_deprecated]
```

Comment:
I suggest a different variable name - Maybe `preferred_providers` or `current_providers`? `default_providers` suggests you're filtering by `is_default == True`, but that's not the case (nor would it work given the current data on DukeDS)